src/tests/test_upgrade_agent.py

```diff
@@ -0,0 +1,189 @@
+    if match is None:
+        return None
+    x = int(match.group("x"))
+    y = int(match.group("y"))
+    z = int(match.group("z"))
+    r = match.group("r")
+    if r is not None:
+        r = int(r)
+    else:
+        r = 0
+    return [x, y, z, r]
+
+
+class TestUpgradeAgent(BaseTest):
+    @JunitTestSuite()
+    def test_upgrade_agent(self, cluster):
+        """
+        This test will prepare the cluster with the previouis agent image. Once it is ready to install it will restart
+        the service with an agent image that doesn't exist, so that all the hosts will move to `insufficient`. Then it
+        will restart the service again with the current image, so that host will upgrade and move to `ready` again.
+        """
+        # Find the namespace:
+        namespace = os.environ.get("NAMESPACE", "assisted-installer")
+
+        # Create the API client:
+        client = kubernetes.dynamic.DynamicClient(
+            kubernetes.client.api_client.ApiClient(
+                configuration=kubernetes.config.load_kube_config(),
+            ),
+        )
+        configmaps_api = client.resources.get(api_version="v1", kind="ConfigMap")
+        deployments_api = client.resources.get(api_version="apps/v1", kind="Deployment")
+
+        # Get the digest and tags of the agent image currently used by the service:
+        service_config = configmaps_api.get(namespace=namespace, name="assisted-service-config")
+        current_image = service_config.data["AGENT_DOCKER_IMAGE"]
+        agent_repo = image_repo(current_image)
+
+        # We expect the repository to contain tags like 'v2.3.1'. We want to find the newest of those tags that is
+        # actually different (different digest) than the current tag used by the service, to make sure that we are using
+        # two different images.
+        all_tags = repo_tags(agent_repo)
+        candidate_pairs = []
+        for tag in all_tags:
+            parsed = parse_tag(tag)
+            if parsed is not None:
+                candidate_pairs.append((tag, parsed))
+        candidate_pairs.sort(key=lambda pair: pair[1], reverse=True)
+        candidate_tags = list(map(lambda pair: pair[0], candidate_pairs))
+        previous_tag = None
+        current_digest = image_digest(current_image)
+        for candidate_tag in candidate_tags:
+            candidate_digest = image_digest(f"{agent_repo}:{candidate_tag}")
+            if candidate_digest != current_digest:
+                previous_tag = candidate_tag
+                break
+        assert previous_tag, "Failed to find previous tag for '{current_image}'"
+        previous_image = f"{agent_repo}:{previous_tag}"
+        broken_image = f"{agent_repo}:broken"
+        log.info(f"Previoius image is '{previous_image}'")
+        log.info(f"Broken image is '{broken_image}'")
+        log.info(f"Current image is '{current_image}'")
+
+        def restart_service(agent_image: str) -> None:
```

Comment:
define it as a class method. it will simplify this test method a bit (``test_upgrade_agent`` is too long right now to get a grasp of the test)

Reply:
It can't just be made a class method because it shares the namespace and Kubernetes API clients with the `test_upgrade_agent` method. I will try to move it out anyhow.

Reply:
Moved it to a separate method.